x/mongo/driver/topology/topology_options.go

```diff
@@ -7,451 +7,370 @@
-	replicaSetName         string
-	seedList               []string
-	serverOpts             []ServerOption
-	cs                     connstring.ConnString // This must not be used for any logic in topology.Topology.
-	uri                    string
-	serverSelectionTimeout time.Duration
-	serverMonitor          *event.ServerMonitor
-	srvMaxHosts            int
-	srvServiceName         string
-	loadBalanced           bool
+// Config is used to construct a topology.
+type Config struct {
+	Mode                   MonitorMode
+	ReplicaSetName         string
+	SeedList               []string
+	ServerOpts             []ServerOption
+	ConnString             connstring.ConnString // This must not be used for any logic in topology.Topology.
+	URI                    string
+	ServerSelectionTimeout time.Duration
+	ServerMonitor          *event.ServerMonitor
+	SRVMaxHosts            int
+	SRVServiceName         string
+	LoadBalanced           bool
 }
 
-func newConfig(opts ...Option) (*config, error) {
-	cfg := &config{
-		seedList:               []string{"localhost:27017"},
-		serverSelectionTimeout: 30 * time.Second,
-	}
+func hasOptionalData(co options.ClientOptions) bool {
+	return co.Direct != nil && *co.Direct ||
+		co.ServerMonitor != nil ||
+		co.ReplicaSet != nil ||
+		co.ServerSelectionTimeout != nil ||
+		co.LoadBalanced != nil
+}
 
-	for _, opt := range opts {
-		if opt == nil {
-			continue
-		}
-		err := opt(cfg)
-		if err != nil {
-			return nil, err
-		}
+// convertToDriverAPIOptions converts a options.ServerAPIOptions instance to a driver.ServerAPIOptions.
+func convertToDriverAPIOptions(s *options.ServerAPIOptions) *driver.ServerAPIOptions {
+	driverOpts := driver.NewServerAPIOptions(string(s.ServerAPIVersion))
+	if s.Strict != nil {
+		driverOpts.SetStrict(*s.Strict)
 	}
-
-	return cfg, nil
+	if s.DeprecationErrors != nil {
+		driverOpts.SetDeprecationErrors(*s.DeprecationErrors)
+	}
+	return driverOpts
 }
 
-// WithConnString configures the topology using the connection string.
-func WithConnString(fn func(connstring.ConnString) connstring.ConnString) Option {
-	return func(c *config) error {
-		cs := fn(c.cs)
-		c.cs = cs
-
-		if cs.ServerSelectionTimeoutSet {
-			c.serverSelectionTimeout = cs.ServerSelectionTimeout
-		}
-
-		var connOpts []ConnectionOption
-
-		if cs.AppName != "" {
-			c.serverOpts = append(c.serverOpts, WithServerAppName(func(string) string { return cs.AppName }))
-		}
-
-		if cs.Connect == connstring.SingleConnect || (cs.DirectConnectionSet && cs.DirectConnection) {
-			c.mode = SingleMode
-		}
+type client interface {
+	GetClusterClock() *session.ClusterClock
+	SetServerAPI(*driver.ServerAPIOptions)
+}
 
-		c.seedList = cs.Hosts
+// newConfig will translate data from client options into a topology config for building non-default deployments.
+func newConfig(co *options.ClientOptions, client client) (*Config, error) {
+	const defaultMaxPoolSize = 100
+	var serverAPI *driver.ServerAPIOptions = nil
```

Comment:
```suggestion
	var serverAPI *driver.ServerAPIOptions
```
Isn't specifying the zero-value here and below redundant?